### ISSUE TITLE
Optional vs Exceptions

### ISSUE THREAD

**Issue by BraisGabin** [PoGoTools/PoGoTools#3]
Move the disccussion from https://github.com/farkam135/GoIV/issues/524#issuecomment-256592958

**Comment by BraisGabin** [comment]
I must say that I'm not familiar with `Either`. I found some information about that class but it's not on [Guava javadoc](https://google.github.io/guava/releases/20.0/api/docs/allclasses-noframe.html). (They just released the v20 🎉).

Anyway, I don't like `Either`. You must ask about which is the correct param and then request it. It's ok if you want to work just with the values and, some times, look the reason. But if you work heavily with the reasons you must write a lot of boilerplate.

I'm ok implementing both interfaces: One using Optional and other using Exceptions. The implementation is really easy:

``` java
Optional<String> candy() {
  try {
    return Optional.of(riskyScreenReader.candy());
  } catch (DeclaredException1 | DeclaredException2 e) {
    return Optiona.absent();
  }
}
```

**Comment by Blaisorblade** [comment]
Yep, that's fine with me — `Either` works best in more functional languages.

BTW, what's your language background or preferred style (object-oriented, functional programming)?
I'm not asking to discuss which language is best, just to take that into account when making future proposals. I've seen you use functional programming idioms in pokescreenshot (like `flatMap` IIRC)—I wouldn't have proposed `Either` otherwise. But then you mentioned "pythonic" recently...

**Comment by BraisGabin** [comment]
I like OO with [reactive programming](http://reactivex.io/). I like some things of Python too, for example: [Easier to ask for forgiveness than permission](https://docs.python.org/2/glossary.html#term-eafp).

And with Exceptions I mean Checked Exceptions that Java forces you to handle. Handle unchecked Exceptions is hell!

**Comment by Blaisorblade** [comment]
Thanks for the clarification. (And yes, Checked Exceptions please).
EDIT: I think we have an agreement on the API you proposed (https://github.com/PoGoTools/PoGoTools/issues/3#issuecomment-257081589). So I guess this can be closed in favor of the next step, #4? I'll leave that to you.

**Comment by harkin** [comment]
If this is going to be a library used in some number of android apps I think depending on Guava isn't ideal. It's forcing that dependency on anyone that wants to use this library and Guava is quite a large one. It's very much up for debate whether you should or shouldn't use Guava in Android apps (google 'android guava' if you have time to burn) for that reason. My feeling is that there's a larger responsibility on a library to be lightweight. [A link](https://github.com/futurice/android-best-practices#methodlimitation) that says some of this

It's possible to provide a public interface that uses `Optional`s without bundling Guava (have it as a compile time only dependency) and only apps that use Guava can use that API then. This approach would require not using `Optional`s internally though so it probably doesn't suit what you're discussing here. 

If `Optional` is the only thing from Guava being used it's a pretty tiny API and could reasonably be moved into this project.

On the flipside this is a relatively small project so getting very stringent about 'best practices' might be quite boring, not fun, and some effort. 

Personally I prefer checked exceptions. Dealing with the ambiguity of 'is something here or not' is not a system I like. You end up having branching logic somewhere anyway to handle both cases. If you fail early when inputs are unexpected/invalid you can be sure past a certain point in the program that you're dealing with real, sanitised data.

**Comment by Blaisorblade** [comment]
> Dealing with the ambiguity of 'is something here or not' is not a system I like.

As IIRC mentioned above, that's not a choice we have because, when scanning a snapshot, scanning of some fields might fail, yet GoIV wants to go on and use the correctly scanned fields. So the choice is between `Optional` and `null`. So dealing with that ambiguity is part of the GoIV requirements. Since the plan is to wrap exceptions into `Optional` at the border, that could be done in GoIV in the worst case.

I don't have time to check myself, but I'd be more sympathetic to the "method count" argument given data showing that GoIV is close to the threshold, or given some other prospective client which might actually run into the problem.

**Comment by harkin** [comment]
I'm mostly playing devils advocate, wanted to add some conversation 👍  `Optional` > `null` for sure

GoIV is no where near it actually. The limit is 64k and GoIV clocks in at a cool 36774 (for the offline debug build).

I can't really think of many features that would instantly push it above or close to the limit. Anything that requires play-services or that family of libraries would, like something location based or push notifications.

**Comment by BraisGabin** [comment]
I must say that I was thinking about 64k problem too. But I didn't want to talk about that. We can try to face this problem later. We can always create a different dependecy with the guava layer if we face some problem. Somthing like:

```
compile com.github.pogotools:pogotools:x.y.z
compile com.github.pogotools:pogotools-guava:x.y.z // just a class with exceptions to Optional
```

Or just drop the support pointing out a gist with the class that you must copy&paste. if you want to use that class.

**Comment by harkin** [comment]
It's probably not going to be a huge issue for anyone. Having two public interface classes in the library, `Xyz` and `XyzGuava` with the former exposing exceptions and the latter wrapping them in `Optional`s might work? If the `Optional` use doesn't extend past the public interface then you can rely on the app to import Guava if it wants to use the nicer interface and it's fine without. That requires a lot of discipline though to make sure there are no accidental uses of Guava-y things in the internals.

It's easier to just depend on Guava I think and tackle it later if it's an issue for anyone
